`stack setup --upgrade-cabal` doesn't work for Cabal-2.2.0.1 + ghc-8.2.2

### Steps to reproduce

`stack --resolver lts-11.11 setup --upgrade-cabal`

### Expected

Upgrades cabal version (I have `Cabal-2.0.1.0` installed, but this attempts to install `Cabal-2.2.0.1`)

### Actual

```
Version 1.8.0, Git revision 33d6eddb065f26ea31936434eff062c60aab1d6a (dirty) (5867 commits) x86_64 hpack-0.28.2
2018-06-09 12:46:14.249903: [debug] Checking for project config at: /home/mgsloan/fpco/test-stack/stack.yaml
@(src/Stack/Config.hs:850:9)
2018-06-09 12:46:14.250118: [debug] Checking for project config at: /home/mgsloan/fpco/stack.yaml
@(src/Stack/Config.hs:850:9)
2018-06-09 12:46:14.250169: [debug] Checking for project config at: /home/mgsloan/stack.yaml
@(src/Stack/Config.hs:850:9)
2018-06-09 12:46:14.250210: [debug] Checking for project config at: /home/stack.yaml
@(src/Stack/Config.hs:850:9)
2018-06-09 12:46:14.250247: [debug] Checking for project config at: /stack.yaml
@(src/Stack/Config.hs:850:9)
2018-06-09 12:46:14.250279: [debug] No project config file found, using defaults.
@(src/Stack/Config.hs:880:13)
2018-06-09 12:46:14.251494: [debug] Using resolver: lts-11.11 specified on command line
@(src/Stack/Config.hs:546:7)
2018-06-09 12:46:14.251627: [debug] Run from outside a project, using implicit global project config
@(src/Stack/Config.hs:570:13)
2018-06-09 12:46:14.251912: [debug] Decoding build plan from: /home/mgsloan/.stack/build-plan/lts-11.11.yaml
@(src/Stack/Snapshot.hs:164:5)
2018-06-09 12:46:14.251967: [debug] Trying to decode /home/mgsloan/.stack/build-plan-cache/lts-11.11.cache
@(src/Stack/Snapshot.hs:156:32)
2018-06-09 12:46:14.257895: [debug] Success decoding /home/mgsloan/.stack/build-plan-cache/lts-11.11.cache
@(src/Stack/Snapshot.hs:156:32)
2018-06-09 12:46:14.258782: [debug] Run process: /sbin/ldconfig -p
@(src/Stack/Setup.hs:580:22)
2018-06-09 12:46:14.260565: [debug] Process finished in 2ms: /sbin/ldconfig -p
@(src/Stack/Setup.hs:580:22)
2018-06-09 12:46:14.260958: [debug] Found shared library libtinfo.so.5 in 'ldconfig -p' output
@(src/Stack/Setup.hs:589:29)
2018-06-09 12:46:14.261421: [debug] Did not find shared library libtinfo.so.6
@(src/Stack/Setup.hs:603:38)
2018-06-09 12:46:14.261502: [debug] Did not find shared library libncursesw.so.6
@(src/Stack/Setup.hs:603:38)
2018-06-09 12:46:14.261558: [debug] Found shared library libgmp.so.10 in 'ldconfig -p' output
@(src/Stack/Setup.hs:589:29)
2018-06-09 12:46:14.261631: [debug] Did not find shared library libgmp.so.3
@(src/Stack/Setup.hs:603:38)
2018-06-09 12:46:14.261665: [debug] Potential GHC builds: standard
@(src/Stack/Setup.hs:632:9)
2018-06-09 12:46:14.261703: [debug] Found already installed GHC builds: standard
@(src/Stack/Setup.hs:458:13)
2018-06-09 12:46:14.262452: [info] Manipulating the global Cabal is only for debugging purposes
@(src/Stack/Setup.hs:688:5)
2018-06-09 12:46:14.262517: [debug] Trying to decode /home/mgsloan/.stack/indices/Hackage/01-index.cache
@(src/Stack/PackageIndex.hs:411:23)
2018-06-09 12:46:14.409854: [debug] Success decoding /home/mgsloan/.stack/indices/Hackage/01-index.cache
@(src/Stack/PackageIndex.hs:411:23)
2018-06-09 12:46:14.412436: [debug] Getting Cabal package version
@(src/Stack/GhcPkg.hs:180:5)
2018-06-09 12:46:14.412542: [debug] Run process: /home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/bin/ghc-pkg --no-user-package-db field --simple-output Cabal version
@(src/Stack/GhcPkg.hs:76:10)
2018-06-09 12:46:14.425545: [debug] Process finished in 13ms: /home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/bin/ghc-pkg --no-user-package-db field --simple-output Cabal version
@(src/Stack/GhcPkg.hs:76:10)
2018-06-09 12:46:14.425780: [info] Installing Cabal-2.2.0.1 to replace 2.0.1.0
@(src/Stack/Setup.hs:722:9)
2018-06-09 12:46:14.624308: [debug] Run process within /tmp/stack-cabal-upgrade13671/Cabal-2.2.0.1/: /home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/bin/ghc Setup.hs
@(src/Stack/Setup.hs:739:43)
[  1 of 216] Compiling Distribution.Compat.Binary ( Distribution/Compat/Binary.hs, Distribution/Compat/Binary.o )
[  2 of 216] Compiling Distribution.Compat.Directory ( Distribution/Compat/Directory.hs, Distribution/Compat/Directory.o )
[  3 of 216] Compiling Distribution.Compat.Exception ( Distribution/Compat/Exception.hs, Distribution/Compat/Exception.o )
[  4 of 216] Compiling Distribution.Compat.Internal.TempFile ( Distribution/Compat/Internal/TempFile.hs, Distribution/Compat/Internal/TempFile.o )
[  5 of 216] Compiling Distribution.Compat.Map.Strict ( Distribution/Compat/Map/Strict.hs, Distribution/Compat/Map/Strict.o )
[  6 of 216] Compiling Distribution.Compat.MonadFail ( Distribution/Compat/MonadFail.hs, Distribution/Compat/MonadFail.o )
[  7 of 216] Compiling Distribution.Compat.Newtype ( Distribution/Compat/Newtype.hs, Distribution/Compat/Newtype.o )
[  8 of 216] Compiling Distribution.Compat.Semigroup ( Distribution/Compat/Semigroup.hs, Distribution/Compat/Semigroup.o )
[  9 of 216] Compiling Distribution.Compat.Stack ( Distribution/Compat/Stack.hs, Distribution/Compat/Stack.o )
[ 10 of 216] Compiling Distribution.Compat.Prelude ( Distribution/Compat/Prelude.hs, Distribution/Compat/Prelude.o )
[ 11 of 216] Compiling Distribution.Compat.ReadP ( Distribution/Compat/ReadP.hs, Distribution/Compat/ReadP.o )
[ 12 of 216] Compiling Distribution.Compat.Parsing ( Distribution/Compat/Parsing.hs, Distribution/Compat/Parsing.o )

Distribution/Compat/Parsing.hs:61:1: error:
    Could not find module ‘Text.Parsec’
    Use -v to see a list of the files searched for.
   |
61 | import qualified Text.Parsec as Parsec
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Received ExitFailure 1 when running
Raw command: /home/mgsloan/.stack/programs/x86_64-linux/ghc-8.2.2/bin/ghc Setup.hs
Run from: /tmp/stack-cabal-upgrade13671/Cabal-2.2.0.1/
```

Side note: I found the message "Manipulating the global Cabal is only for debugging purposes" to be surprising.  I find the phrasing to be quite humorous, but this does not match with my understanding of `--upgrade-cabal`.  It was added in https://github.com/commercialhaskell/stack/commit/90ae9ece050a9fe06bc317a5573ca58111bb5d31#diff-4968f87e674615592e572f2ec593e5b0R636

### Stack version

```
$ stack --version
Version 1.8.0, Git revision 33d6eddb065f26ea31936434eff062c60aab1d6a (dirty) (5867 commits) x86_64 hpack-0.28.2
```

### Method of installation

Installed via `stack install` in latest git repo.

--- a/ChangeLog.md
+++ b/ChangeLog.md
@@ -14,6 +14,12 @@ Behavior changes:
 * `stack build` will now announce when sublibraries of a package are being
   build, in the same way executables, tests, benchmarks and libraries are
   announced
+* The `--upgrade-cabal` option to `stack setup` has been
+  deprecated. This feature no longer works with GHC 8.2 and
+  later. Furthermore, the reason for this flag originally being
+  implemented was drastically lessened once Stack started using the
+  snapshot's `Cabal` library for custom setups. See:
+  [#4070](https://github.com/commercialhaskell/stack/issues/4070).
 
 Other enhancements:
 

--- a/doc/GUIDE.md
+++ b/doc/GUIDE.md
@@ -1787,10 +1787,6 @@ users. Here's a quick rundown:
 * `stack upgrade` will build a new version of stack from source.
     * `--git` is a convenient way to get the most recent version from master for
       those testing and living on the bleeding edge.
-* `stack setup --upgrade-cabal` can install a newer version of the Cabal
-  library, used for performing actual builds. You shouldn't generally do this,
-  since new Cabal versions may introduce incompatibilities with package sets,
-  but it can be useful if you're trying to test a specific bugfix.
 * `stack ls snapshots` will list all the local snapshots by
   default. You can also view the remote snapshots using `stack ls
   snapshots remote`. It also supports option for viewing only lts

--- a/src/Stack/Setup.hs
+++ b/src/Stack/Setup.hs
@@ -684,7 +684,8 @@ upgradeCabal :: (HasConfig env, HasGHCVariant env)
              -> UpgradeTo
              -> RIO env ()
 upgradeCabal wc upgradeTo = do
-    logInfo "Manipulating the global Cabal is only for debugging purposes"
+    logWarn "Using deprecated --upgrade-cabal feature, this is not recommended"
+    logWarn "Manipulating the global Cabal is only for debugging purposes"
     let name = $(mkPackageName "Cabal")
     rmap <- resolvePackages Nothing mempty (Set.singleton name)
     installed <- getCabalPkgVer wc
@@ -717,6 +718,10 @@ doCabalInstall :: (HasConfig env, HasGHCVariant env)
                -> Version
                -> RIO env ()
 doCabalInstall wc installed wantedVersion = do
+    when (wantedVersion >= $(mkVersion "2.2")) $ do
+        logWarn "--upgrade-cabal will almost certainly fail for Cabal 2.2 or later"
+        logWarn "See: https://github.com/commercialhaskell/stack/issues/4070"
+        logWarn "Valiantly attempting to build it anyway, but I know this is doomed"
     withSystemTempDir "stack-cabal-upgrade" $ \tmpdir -> do
         logInfo $
             "Installing Cabal-" <>

--- a/src/Stack/SetupCmd.hs
+++ b/src/Stack/SetupCmd.hs
@@ -63,7 +63,7 @@ cabalUpgradeParser = Specific <$> version' <|> latestParser
          <> OA.help "Install a specific version of Cabal" )
         latestParser = OA.flag' Latest (
             OA.long "upgrade-cabal"
-         <> OA.help "Install latest version of Cabal globally" )
+         <> OA.help "DEPRECATED Install latest version of Cabal globally" )
 
 setupParser :: OA.Parser SetupCmdOpts
 setupParser = SetupCmdOpts

